packages/ai-jsx/src/lib/openai.tsx

```diff
@@ -438,59 +422,97 @@ export async function* OpenAIChatModel(
     }
   >;
 
-  const currentMessage = { content: undefined, function_call: undefined } as Partial<ChatCompletionResponseMessage>;
-  let finishReason: string | undefined = undefined;
-  for await (const deltaMessage of openAiEventsToJson<ChatCompletionDelta>(
-    asyncIteratorOfFetchStream(chatResponse.body!.getReader())
-  )) {
-    logger.trace({ deltaMessage }, 'Got delta message');
-    finishReason = finishReason ?? deltaMessage.choices[0].finish_reason;
-    const delta = deltaMessage.choices[0].delta;
-    if (delta.role) {
-      currentMessage.role = deltaMessage.choices[0].delta.role;
+  const iterator = openAiEventsToJson<ChatCompletionDelta>(asyncIteratorOfFetchStream(chatResponse.body!.getReader()))[
+    Symbol.asyncIterator
+  ]();
+
+  // We have a single response iterator, but we'll wrap tokens _within_ the structure of <AssistantMessage> or <FunctionCall>
+  // components. This:
+  //  - Allows our stream to be append-only and therefore eagerly rendered in append-only contexts.
+  //  - Preserves the output structure to allow callers to extract/separate <AssistantMessage> and <FunctionCall> messages.
+  //  - Allows the intermediate states of the stream to include "partial" <FunctionCall> elements with healed JSON.
+  //
+  // This requires some gymnastics because several components will share a single iterator that can only be consumed once.
+  // That is, the logical loop execution is spread over multiple functions (closures over the shared iterator).
+  async function advance(): Promise<Partial<ChatCompletionResponseMessage> | null> {
+    const next = await iterator.next();
+    if (next.done) {
+      return null;
     }
-    if (delta.content) {
-      currentMessage.content = currentMessage.content ?? '';
-      currentMessage.content += delta.content;
-      if (!props.experimental_streamFunctionCallOnly) {
-        yield delta.content;
-      }
+
+    logger.trace({ deltaMessage: next.value }, 'Got delta message');
+    return next.value.choices[0].delta;
+  }
+
+  let delta = await advance();
+  while (delta !== null) {
+    if (delta.role === 'assistant') {
+      // Memoize the stream to ensure it renders only once.
+      const assistantStream = memo(
+        (async function* (): AI.RenderableStream {
+          yield AI.AppendOnlyStream;
+
+          while (delta !== null) {
+            if (delta.content) {
+              yield delta.content;
+            }
+            if (delta.function_call) {
+              break;
+            }
+            delta = await advance();
```

Comment:
I'm confused. So the `advance` function is shared between the two streams, plus the scope above? It's the same data stream, right? So how do we now which one calls it when? How is there no race-condition? Like if you `await` it and it turns out to be a `function_call`, doesn't that mean the other streams are "not" going to get this `delta` and we just threw out the data that was inside that `delta`?

Reply:
The (dark) magic is the `memo`/`yield`/`await render` sequence which acts like a memory barrier in that the code below it will not execute until the generators above it have been flushed once. (This ensures that the execution-order of the code is top-down.)

There are two side-effects these closure-generators have:
- Calling `advance` mutates the shared iterator
- Assigning to `delta`, which is a shared `let` binding. (Note that each closure-generator starts by looking at the existing value of `delta`, rather than starting with an `advance`.)